### PR TITLE
feat(gatsby): Remove old touchNode signature

### PR DESCRIPTION
**Pull request by LekoArts**
## Description

Similarly to https://github.com/gatsbyjs/gatsby/pull/29205 this removes an old arg signature, namely calling `touchNode` with a string (the id). This PR also adds a new deprecation so that in the future `touchNode` will be called like `createNode` and `deleteNode` -- with `touchNode(node)`. 

### Documentation

The docs (source plugin guide) are updated.

## Related Issues

[ch23714]

### CODE REVIEW

File: packages/gatsby/src/redux/actions/public.js [hunk 1]
```diff
@@ -836,36 +836,39 @@ actions.createNode = (...args) => dispatch => {
  * nodes from a remote system that can return only nodes that have
  * updated. The source plugin then touches all the nodes that haven't
  * updated but still exist so Gatsby knows to keep them.
- * @param {Object} $0
- * @param {string} $0.nodeId The id of a node
+ * @param {object} node A node object. See the "createNode" action for more information about the node object details.
```
Comment (gillkyle) [inline packages/gatsby/src/redux/actions/public.js:839]:
```suggestion
 * @param {Object} node A node object. See the "createNode" action for more information about the node object details.
```

I'm not sure it matters, but I think we capitalize this elsewhere and it looks like [JSDoc](https://jsdoc.app/tags-param.html#names-types-and-descriptions) capitalizes `Object` in their docs? 🤷‍♂️

**Review by gillkyle — DISMISSED** [review]
This looks good to me, the warning looks helpful too!

I left one small suggestion that I'm not even sure makes a difference is all 🙂

**Comment by LekoArts** [comment]
@TylerBarnes I also had to update the WordPress plugin (will do the same in https://github.com/gatsbyjs/gatsby/pull/29205)

**Comment by TylerBarnes** [comment]
@LekoArts does this depend on a newer version of Gatsby? I haven't PR'd over the tests from the WP experimental repo yet so I'm a bit nervous to make changes to it.

**Comment by LekoArts** [comment]
@TylerBarnes The change is made only here in this PR, so how about you merging your tests, I rebase onto master and we'll see if they work?

**Comment by LekoArts** [comment]
@TylerBarnes To get myself unblocked and you not urged to merge your tests ASAP I reverted the WordPress changes :) The current state is now using deprecated APIs then, but it still works.

File: packages/gatsby/src/redux/actions/public.js [hunk 2]
```diff
@@ -836,36 +836,39 @@ actions.createNode = (...args) => dispatch => {
...
+    typeOwners[internalNode.internal.type] = internalNode.internal.owner
   }
 
   return {
     type: `TOUCH_NODE`,
     plugin,
-    payload: nodeId,
+    payload: id,
   }
 }
```
Comment (wardpeet) [inline packages/gatsby/src/redux/actions/public.js:873]: We're doing double getNode lookup. We should do node lookup when old behavior is called.

```js
actions.touchNode = (node: any, plugin?: Plugin) => {
  // TODO(v4): Remove this deprecation warning and only allow touchNode(node)
  if (node && node.nodeId) {
    let msg =
      `Calling "touchNode" with an object containing the nodeId is deprecated. Please pass ` +
      `the node directly to the function: touchNode(node)`


    if (plugin && plugin.name) {
      msg = msg + ` "touchNode" was called by ${plugin.name}`
    }


    report.warn(msg)


    node = getNode(node.nodeId)
  }

  if (node && !typeOwners[node.internal.type]) {
    typeOwners[node.internal.type] = node.internal.owner
  }

  return {
    type: `TOUCH_NODE`,
    plugin,
    payload: node.id,
  }
}
```

**Review by wardpeet — APPROVED** [review]

**Pull request merged**